# Black Scholes Vanilla European Call Prices

## Imports

In [1]:
import datetime
y=datetime.datetime.now()
print(y.strftime("%H:%M:%S:%f %A %d %B %Y %z" ))

14:21:07:326164 Sunday 23 December 2018 


In [2]:
import os
print("Current working directory",os.getcwd())

Current working directory C:\Users\Peter\Documents\code\lbs\learningBlackScholes


In [3]:
import numpy as np
np.__version__

'1.15.4'

In [4]:
import pandas as pd
pd.__version__

'0.23.4'

In [5]:
import matplotlib
import matplotlib.pyplot as plt

In [6]:
from math import log, sqrt, exp
from scipy import stats

In [7]:
import decimal
from decimal import Decimal
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [8]:
decimal.getcontext().prec=112
#increase the number precision

## Functions

In [9]:
'''
Black-Scholes vanilla European call price function, with the default values as shown in the example in Hull (2009)
'''

def bs(S=42,K=40,r=0.1,sigma=0.2,T=0.5):
    if ( np.any(S)==0 ):
        return 0
    elif (np.any(K)==0):
        return S    
    else:    
        d1 = (np.log(S/K)  +   (r+(np.square(sigma)/2) )* T )  / (sigma*np.sqrt(T))  
        d2 = d1 - (sigma * np.sqrt(T) )   
        c = S*stats.norm.cdf(d1, 0.0, 1.0)-  K*np.exp(-r*T)*stats.norm.cdf(d2, 0.0, 1.0)
        return c  

In [10]:
'''
Black-Scholes vanilla European call price function, with the default values as shown in the example in Hull (2009)
'''

def bsDecimal(S=42,K=40,r=0.1,sigma=0.2,T=0.5):
    
    S=Decimal(S)
    K=Decimal(K)
    r=Decimal(r)
    sigma=Decimal(sigma)
    T=Decimal(T)
    
    if ( np.any(S)==0 ):
        return 0
    elif (np.any(K)==0):
        return S    
    else:    
        d1 = (  ( Decimal (S/K) ).ln()  +   Decimal ( (r+(np.square(sigma)/2) )* T ) )  / Decimal ( (sigma*np.sqrt(T)) ) 
        d2 = d1 - Decimal ( (sigma * np.sqrt(T) )   )
        
        '''
        print(d1)
        print(round(d1,20))
        print(float(round(d1,20)))
        print()
        print(d2)
        print(round(d2,20))
        print(float(round(d2,20)))
        '''
        
        #c = Decimal( S* Decimal( stats.norm.cdf(d1, 0.0, 1.0) )  )\
        #            - Decimal (K * (  np.exp(-r*T)*Decimal(stats.norm.cdf(d2, 0.0, 1.0))) )
        
        round_d1=float(round(d1,20))
        print(type(round_d1))
        round_d2=float(round(d2,20))
        print(type(round_d1))
        
        cdf_d1=Decimal ( stats.norm.cdf(round_d1, 0.0, 1.0) ) 
        cdf_d2=Decimal ( stats.norm.cdf(round_d2, 0.0, 1.0) )
        
        c = Decimal (
        
        (
            S*cdf_d1 
        )
            -
        (
            K*np.exp(-r*T)*cdf_d2
        )
        )
        
        return c  

In [11]:
x=Decimal(42)

In [12]:
bsDecimal(S=x)

<class 'float'>
<class 'float'>


Decimal('4.75942239287153143007131791898744713680478009470034327250796964568769722921726499194780886164081340434109854130')

In [13]:
y=float(10)

In [14]:
x+Decimal(y)

Decimal('52')

In [15]:
type(y)

float

In [16]:
bs(S=42)

4.759422392871532

### Function bs() typically produces the Euro call value of 4.759422392871532

In [17]:
bs(K=0)

42

In [18]:
type(y)

float

Hull's value for the default is 4.759, same as the output of the bs function with the default parameters

In [19]:
'''
BSM Call Function taken from Hilpisch (2014), Chapter 10
'''
def bsm_call_value(S0, K, T, r, sigma):
    S0 = float(S0)
    d1 = (log(S0 / K) + (r + 0.5 * sigma ** 2) * T) / (sigma * sqrt(T))
    d2 = (log(S0 / K) + (r - 0.5 * sigma ** 2) * T) / (sigma * sqrt(T))
    value = (S0 * stats.norm.cdf(d1, 0.0, 1.0)
             - K * exp(-r * T) * stats.norm.cdf(d2, 0.0, 1.0))
    return value

# stats.norm.cdf --> cumulative distribution function
#                    for normal distribution

In [20]:
bsm_call_value(42,40,0.5,0.1,0.2)

4.759422392871532

## Parameters for the model

In [21]:
NUMBER_S=10

In [22]:
INCREMENT_S=( 1/1 ) 

In [23]:
INCREMENT_S

1.0

In [24]:
NUMBER_X=10

In [25]:
INCREMENT_X=1/1

In [26]:
INCREMENT_X

1.0

In [27]:
NUMBER_SIGMA=10

In [28]:
INCREMENT_SIGMA=1/1

In [29]:
INCREMENT_SIGMA

1.0

In [30]:
NUMBER_T=252

In [31]:
INCREMENT_T=1

In [32]:
INCREMENT_T

1

In [61]:
optionValues=np.zeros(NUMBER_S*NUMBER_X*NUMBER_SIGMA)

In [62]:
optionValues.shape

(1000,)

In [33]:
for m in np.arange (1, 26, 0.1):
    print("{:1.1f} \t {:3.6f}".format(m,bs(m)))    

1.0 	 0.000000
1.1 	 0.000000
1.2 	 0.000000
1.3 	 0.000000
1.4 	 0.000000
1.5 	 0.000000
1.6 	 0.000000
1.7 	 0.000000
1.8 	 0.000000
1.9 	 0.000000
2.0 	 0.000000
2.1 	 0.000000
2.2 	 0.000000
2.3 	 0.000000
2.4 	 0.000000
2.5 	 0.000000
2.6 	 0.000000
2.7 	 0.000000
2.8 	 0.000000
2.9 	 0.000000
3.0 	 0.000000
3.1 	 0.000000
3.2 	 0.000000
3.3 	 0.000000
3.4 	 0.000000
3.5 	 0.000000
3.6 	 0.000000
3.7 	 0.000000
3.8 	 0.000000
3.9 	 0.000000
4.0 	 0.000000
4.1 	 0.000000
4.2 	 0.000000
4.3 	 0.000000
4.4 	 0.000000
4.5 	 0.000000
4.6 	 0.000000
4.7 	 0.000000
4.8 	 0.000000
4.9 	 0.000000
5.0 	 0.000000
5.1 	 0.000000
5.2 	 0.000000
5.3 	 0.000000
5.4 	 0.000000
5.5 	 0.000000
5.6 	 0.000000
5.7 	 0.000000
5.8 	 0.000000
5.9 	 0.000000
6.0 	 0.000000
6.1 	 0.000000
6.2 	 0.000000
6.3 	 0.000000
6.4 	 0.000000
6.5 	 0.000000
6.6 	 0.000000
6.7 	 0.000000
6.8 	 0.000000
6.9 	 0.000000
7.0 	 0.000000
7.1 	 0.000000
7.2 	 0.000000
7.3 	 0.000000
7.4 	 0.000000
7.5 	 0.000000
7.6 	 0.00

In [34]:
optionS=np.array(np.arange(1,NUMBER_S,INCREMENT_S))

In [35]:
optionX=np.array(np.arange(1,NUMBER_X,INCREMENT_X))

In [36]:
optionSIGMA=np.array(np.arange(1,NUMBER_SIGMA,INCREMENT_SIGMA))

In [37]:
optionT=np.array(np.arange(1,NUMBER_T,INCREMENT_T))

In [38]:
optionS

array([1., 2., 3., 4., 5., 6., 7., 8., 9.])

In [39]:
print("{:2.100}".format(optionS[6]/3))

2.333333333333333481363069950020872056484222412109375


In [40]:
def myFunction(x):
    return (1+ ( 2*x) ) b

In [73]:
counter=0
for s in optionS:
    for x in optionX:
        for sig in optionSIGMA:
            optionValues[counter]=bs(S=s,K=x,sigma=sig,T=0.5)
            counter+=1    


In [76]:
for i in optionValues:
    print(i,'\n')


0.2946192953639579 

0.5324604514967268 

0.7183298242823684 

0.8465998401577595 

0.9248090670100526 

0.9669437654180822 

0.9870012822642098 

0.9954379077476608 

0.9985734359491203 

0.09262822181756644 

0.3662868159208478 

0.6114196787775901 

0.7865890403144296 

0.8948972509069191 

0.9536570385000317 

0.9817406670629263 

0.9935827947333686 

0.9979913367122819 

0.03478942733947489 

0.27589729975158717 

0.5432676376489908 

0.7456654179726583 

0.8737544384784903 

0.944060502687917 

0.9778872638934267 

0.9922106796927016 

0.9975577814002827 

0.014917126307808413 

0.2185654409405362 

0.4937054687976736 

0.7142069462709579 

0.8570277604474009 

0.9363380369105331 

0.974752142702823 

0.991085912372037 

0.9972004684057604 

0.007057051163399289 

0.17896274932806758 

0.45513340624423515 

0.6885380936828593 

0.8430452561605876 

0.9297901275410383 

0.9720695283323446 

0.9901174933653869 

0.9968914589534004 

0.003600659479886713 

0.15004339701292663 

0.42

6.0977776614194035 

6.557014740782368 

6.805044563844164 

6.923283702338457 

6.973062043555318 

6.991573506262241 

2.471393508237072 

3.9823181256739284 

5.178888373835982 

6.00740738638501 

6.513248653944151 

6.785948968327009 

6.915812837306182 

6.9704493953645645 

6.990758660054775 

2.0623350675477057 

3.7272231604770876 

5.028308769976578 

5.926198881104317 

6.4736634690703685 

6.768606357926576 

6.909008975849469 

6.968065354233627 

6.990014051643842 

1.7271055907650088 

3.502884324084865 

4.8927441909335485 

5.852263559298365 

6.43739409084455 

6.752653271296531 

6.902733652521498 

6.965862425086728 

6.98932508056562 

1.452087828857488 

3.303667854921653 

4.769447790017994 

5.7842714652225595 

6.403834325862853 

6.737835705498333 

6.896890257922507 

6.963807504648134 

6.988681577852424 

7.0494486257791475 

7.143037661551113 

7.39005485941662 

7.637287347903439 

7.8128913933237385 

7.915049562907682 

7.965866860999454 

7.98783796011

In [42]:
#optionValues=bs(optionS)

In [43]:
myFunction(optionS)

array([ 3.,  5.,  7.,  9., 11., 13., 15., 17., 19.])

In [57]:
optionValues

array(1000)

In [ ]:
optionS.shape

In [ ]:
optionValues.shape

In [ ]:
optionArray=np.array((optionS,optionValues))

In [ ]:
optionArray.shape

In [ ]:
optionArray=optionArray.T

In [ ]:
type(optionArray)

In [ ]:
optionDF=pd.DataFrame(data=optionArray,columns=["S","Value"])

In [ ]:
optionDF['S']

In [ ]:
optionDF['Value']

In [ ]:
optionDF.shape

In [ ]:
underlying=(42/INCREMENT_S)

In [ ]:
lower=int(underlying-(10/INCREMENT_S))

In [ ]:
underlying

In [ ]:
upper=int(underlying+(10/INCREMENT_S))

In [ ]:
5/INCREMENT_S

In [ ]:
upper

In [ ]:
lower

In [ ]:
optionDF[optionDF['S']==1.0]

In [ ]:
optionDF

In [ ]:
plt.rcParams['figure.figsize'] = [7, 7] # This controls the picture size
fig, ax = plt.subplots()
ax.plot(optionDF['S'][lower:upper],optionDF['Value'][lower:upper])
#ax.plot()
ax.set(xlabel='underlying price', ylabel='option value',
       title='Call option value as function of current price')
ax.grid()

fig.savefig("test.png")
plt.show()

In [ ]:
x=(optionDF['S'][lower:upper])

In [ ]:
y=(optionDF['Value'][lower:upper])

In [ ]:
plt.plot(x,y)

In [ ]:
#optionDF.loc[['S']==42.0 ] #    [(S=37):(S=47)]

optionDF.loc[round(optionDF['S'],5)==42]

In [ ]:
optionDF.loc[optionDF['S']==1]

In [ ]:
optionDF.index

In [ ]:
optionDF.columns

In [ ]:
optionDF.iloc[41000]['Value']

In [ ]:
optionDF.iloc[41000]['S']

In [ ]:
optionDF.loc[optionDF.S==1]

In [ ]:
optionDF['Value']

In [ ]:
optionDF.loc[optionDF[S>41.000]]['Value']

In [ ]:
optionDF.dtypes

In [ ]:
optionDF